src/main/java/com/minecolonies/coremod/placementhandlers/PlacementHandlers.java

```diff
@@ -0,0 +1,394 @@
+            }
+
+            final EnumFacing facing = blockState.getValue(BlockBed.FACING);
+
+            //Set other part of the bed, to the opposite PartType
+            if (blockState.getValue(BlockBed.PART) == BlockBed.EnumPartType.FOOT)
+            {
+                placer.handleBuildingOverBlock(pos);
+                //pos.offset(facing) will get the other part of the bed
+                world.setBlockState(pos.offset(facing), blockState.withProperty(BlockBed.PART, BlockBed.EnumPartType.HEAD), 0x03);
+                world.setBlockState(pos, blockState.withProperty(BlockBed.PART, BlockBed.EnumPartType.FOOT), 0x03);
+                return blockState;
+            }
+            return ActionProcessingResult.ACCEPT;
+        }
+    }
+
+    public static class DoublePlantPlacementHandler implements IPlacementHandler
+    {
+        @Override
+        public Object handle(
+                @NotNull final World world, @NotNull final BlockPos pos, @NotNull final IBlockState blockState,
+                @Nullable final AbstractEntityAIStructure placer)
+        {
+            if (!(blockState.getBlock() instanceof BlockDoublePlant))
+            {
+                return ActionProcessingResult.IGNORE;
+            }
+
+            if (!Configurations.builderInfiniteResources && blockState.getValue(BlockDoublePlant.HALF).equals(BlockDoublePlant.EnumBlockHalf.LOWER)
+                    && placer.checkOrRequestItems(placer.getTotalAmount(BlockUtils.getItemStackFromBlockState(blockState))))
+            {
+                return ActionProcessingResult.DENY;
+            }
+
+            if (blockState.getValue(BlockDoublePlant.HALF).equals(BlockDoublePlant.EnumBlockHalf.LOWER))
+            {
+                placer.handleBuildingOverBlock(pos);
+                world.setBlockState(pos, blockState.withProperty(BlockDoublePlant.HALF, BlockDoublePlant.EnumBlockHalf.LOWER), 0x03);
+                world.setBlockState(pos.up(), blockState.withProperty(BlockDoublePlant.HALF, BlockDoublePlant.EnumBlockHalf.UPPER), 0x03);
+                return blockState;
+            }
+            return ActionProcessingResult.ACCEPT;
+        }
+    }
+
+    public static class SpecialBlockPlacementAttemptHandler implements IPlacementHandler
+    {
+        @Override
+        public Object handle(
+                @NotNull final World world, @NotNull final BlockPos pos, @NotNull final IBlockState blockState,
+                @Nullable final AbstractEntityAIStructure placer)
+        {
+            if (blockState instanceof BlockEndPortal
+                    || blockState instanceof BlockMobSpawner
+                    || blockState instanceof BlockDragonEgg
+                    || blockState instanceof BlockPortal)
+            {
+                return ActionProcessingResult.ACCEPT;
+            }
+            return ActionProcessingResult.IGNORE;
+        }
+    }
+
+    public static class FlowerPotPlacementHandler implements IPlacementHandler
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Reduce this class from 31 to the maximum allowed 30 or externalize it in a public class. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)